File: 백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.py
N = int(input()) # 1 <= N <= 1000
arr = list(map(int, input().split())) # 1 <= x <= 1000

dp = [1] * N


for i in range(N):
    for j in range(i):
        # 순차적으로 선택
        if arr[j] < arr[i]:
            dp[i] = max(dp[i], dp[j] + 1)

print(max(dp))

File: 백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.cc
#include <bits/stdc++.h>
using namespace std;
#define DEBUG true;

int N;
vector<int> arr;
vector<int> dp;


int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    cin >> N;
    arr.resize(N, 0);
    dp.resize(N, 1);
    for (int i = 0 ; i < N ; ++i) cin >> arr[i];
    for (int i = 1 ; i < N ; ++i)
        for (int j = 0 ; j < i ; ++j)
            if (arr[j] < arr[i])
                dp[i] = max(dp[i], dp[j] + 1);

    cout << *max_element(dp.begin(), dp.end());
}
